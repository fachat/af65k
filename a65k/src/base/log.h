/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef LOG_H
#define LOG_H

void log_errno(const char *msg, ...);

void log_error(const char *msg, ...);

#ifdef DEBUG

void log_warn(const char *msg, ...);

void log_info(const char *msg, ...);

void log_debug(const char *msg, ...);

void log_term(const char *msg);

#define	log_rv(rv)	log_error("ERROR RETURN: %d\n", (rv))

#else

#define	log_warn(msg, ...)
#define	log_info(msg, ...)
#define	log_debug(msg, ...)
#define	log_term(msg, ...)
#define	log_rv(rv)

#endif

#endif

