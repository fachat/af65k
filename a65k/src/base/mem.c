/****************************************************************************

    65k processor assembler
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

/*
 * basically a malloc/free wrapper
 */

#include <stdio.h>
#include <malloc.h>
#include <string.h>

#include "types.h"

void mem_module_init(void) {
}


// allocate memory and copy given string
char *mem_alloc_str(const char *orig) {

	int len = strlen(orig);

	char *ptr = malloc(len+1);
	
	strcpy(ptr, orig);

	return ptr;
}

// allocate memory and copy given string, up to number of chars given
char *mem_alloc_strn(const char *orig, int nchars) {

	int len = strlen(orig);
	
	if (len < nchars) {
		nchars = len;
	}

	char *ptr = malloc(nchars+1);
	
	strncpy(ptr, orig, nchars);

	ptr[nchars] = 0;

	return ptr;
}


void *mem_alloc(const type_t *type) {
	// for now just malloc()

	return malloc(type->sizeoftype);
}

void *mem_alloc_c(size_t n, const char *name) {
	// for now just malloc()
	(void) name;

	return malloc(n);
}

// NOTE: does not handle padding, this must be fixed in the type->sizeofstruct value!
void *mem_alloc_n(const size_t n, const type_t *type) {
	// for now just malloc()

	return calloc(n, type->sizeoftype);
}

// NOTE: does not handle padding, this must be fixed in the type->sizeofstruct value!
void *mem_realloc_n(const size_t n, const type_t *type, void *data) {
	// for now just malloc()

	return realloc(data, n * type->sizeoftype);
}

void mem_free(void *ptr) {
	
	free(ptr);
}



