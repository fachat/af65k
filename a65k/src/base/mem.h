/****************************************************************************

    65k processor assembler
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#ifndef MEM_H
#define MEM_H

#include <stdio.h>

#include "types.h"

/*
 * basically a malloc/free wrapper
 */

void mem_module_init(void);

void mem_free(void *ptr);

// alloc single object
void *mem_alloc(const type_t *type);

// alloc multiple object, returning a pointer to an array
void *mem_alloc_n(size_t n, const type_t *type);

// re-alloc multiple objects, returning a pointer to the new array
void *mem_realloc_n(size_t n, const type_t *type, void *orig);

// alloc multiple object, returning a pointer to an array
void *mem_alloc_c(size_t n, const char *name);

// allocate memory and copy given string
char *mem_alloc_str(const char *orig);

// allocate memory and copy given string, up to given number of chars;
// Note: result will always be null byte terminated.
char *mem_alloc_strn(const char *orig, int nchars);

#endif

