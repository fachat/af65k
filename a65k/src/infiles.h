/****************************************************************************

    input file handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef INFILES_H
#define INFILES_H

struct openfile {
        char            *filename;      // the file name part
        char            *filepath;      // the path part
        int             current_line;
        FILE            *filep;
        int             buffer_size;
        char            *buffer;
};
typedef struct openfile openfile_t;

typedef struct {
        const openfile_t        *file;
        int                     lineno;
} position_t;

typedef struct {
	const char 		*line;
	position_t		*position;
} line_t;

void infiles_module_init(void);

// add an include directory to the input file processing
void infiles_includedir(const char *filename);

// register a top level input file 
void infiles_register(const char *filename);

// read a line from the input. Note the result is static and must not be freed.
// Ownership of the actual line stays with the infile module, so line must be copied
// when a modifyable version, or one that exists longer than the next call to infile_readline
// is needed. The position_t pointed to in the line is transferred to the caller
// and should/can be freed
line_t *infiles_readline();

// during read operation, and after parsing, an "include" operation can occur. 
// The indiles_include call opens the file given, and reads from there until end of file,
// before returning to the original file.
void infiles_include(const char *filename);

#endif

