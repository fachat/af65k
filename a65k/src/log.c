/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_warn(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("WRN:");
        vprintf(msg, args);
}

void log_error(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("ERR:");
        vprintf(msg, args);
}

void log_info(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("INF:");
        vprintf(msg, args);
}

void log_debug(const char *msg, ...) {
       va_list args;
       va_start(args, msg);

       printf("DBG:");
        vprintf(msg, args);
}



