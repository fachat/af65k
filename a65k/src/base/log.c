/****************************************************************************

    generic logging interface
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <strings.h>
#include <stdarg.h>
#include <ctype.h>

#include "infiles.h"
#include "log.h"


static err_level level = LEV_WARN;

void log_module_init(err_level l) {
	level = l;
}

void log_set_level(err_level l) {
	level = l;
}


void log_term(const char *msg) {
	
	int newline = 0;

	printf(">>>: ");
	char c;
	while ((c = *msg) != 0) {
		if (newline) {
			printf("\n>>>: ");
		}
		if (isprint(c)) {
			putchar(c);
			newline = 0;
		} else
		if (c == 10) {
			newline = 1;
		} else
		if (c == 13) {
			// silently ignore
		} else {
			printf("{%02x}", ((int)c) & 0xff);
			newline = 0;
		}
		msg++;
	}
	printf("\n");
}

void log_errno(const char *msg, ...) {
       va_list args;
       va_start(args, msg);
	char buffer[1024];

	if (index(msg, '%') != NULL) {
		// the msg contains parameter, so we need to fix it
		vsprintf(buffer, msg, args);
		msg = buffer;
	}
        printf(">> %s: errno=%d: %s\n", msg, errno, strerror(errno));
}

void log_x(err_level lev, const char *msg, ...) {

	const char *prefix = ">>>";
	switch (lev) {
	case LEV_TRACE: prefix = "TRC:"; break;
	case LEV_DEBUG: prefix = "DBG:"; break;
	case LEV_INFO: 	prefix = "INF:"; break;
	case LEV_WARN: 	prefix = "WRN:"; break;
	case LEV_ERROR: prefix = "ERR:"; break;
	case LEV_FATAL: prefix = "FTL:"; break;
	}

	if (lev >= level) {
       		va_list args;
       		va_start(args, msg);

       		printf(prefix);
       		vprintf(msg, args);
		printf("\n");
	}
}




