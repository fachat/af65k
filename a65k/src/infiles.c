/****************************************************************************

    input file handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/

#include <string.h>
#include <strings.h>

#include "log.h"
#include "mem.h"
#include "infiles.h"
#include "linked_list.h"

#define	PATH_SEPARATOR_CHAR 	'/'
#define	PATH_SEPARATOR_STR 	"/"

#define	INITIAL_BUFFER_SIZE	4096



static type_t openfile_memtype = {
	"openfile",
	sizeof(openfile_t)
};

static type_t position_memtype = {
	"position_t",
	sizeof(position_t)
};

static openfile_t *current_file = NULL;

static line_t line;

// contains the stack of open files
// with entries of openfile_t. Note that entries are not 
// deleted (freed) when the file is closed, as other parts
// hold references to e.g. the file name
static list_t *filestack;

// list of input files as given by the command line (char*)
static list_t *infiles;
// iterator over the input files
static list_iterator_t *infile_iter;

// list of include dirs to be used when finding an input file,
// be it from the command line or from include statements
static list_t *incdirs;

void infiles_module_init(void) {

	infiles = linked_list_init();
	infile_iter = NULL;

	incdirs = linked_list_init();

	filestack = linked_list_init();

	current_file = NULL;
}

// add an include directory to the input file processing
void infiles_includedir(const char *dirname) {

	list_add(incdirs, (char*) dirname);
}


// register a top level input file 
void infiles_register(const char *filename) {

	list_add(infiles, (char*) filename);
}

// combine the (optional) path and the name into a full file path,
// allocating memory for it and copying it there.
static char *alloc_fullpath(const char *path, const char *name) {
	
	char *fp = NULL;
	int len = strlen(name);
	int plen = path == NULL ? 0 : strlen(path);

	// two strings, path separator, and final zero byte
	fp = mem_alloc_c(plen + 1 + len + 1, "fullpath");

	fp[0] = '\0';
	if (path != NULL) {
		strcat(fp, path);
		strcat(fp, PATH_SEPARATOR_STR);
	}
	strcat(fp, name);
	
	return fp;
}

// low level open given single path. Return NULL on not found
static openfile_t *open_file(char *filepath) {

	FILE *filep = fopen(filepath, "r");

	if (filep != NULL) {

		openfile_t *of = mem_alloc(&openfile_memtype);
		
		of->buffer_size = INITIAL_BUFFER_SIZE;
		of->buffer = mem_alloc_c(of->buffer_size, "openfile_buffer");

		of->filep = filep;
		of->current_line = 1;
		
		// separate into file name and path
		char *lastsep = rindex(filepath, PATH_SEPARATOR_CHAR);
		if (lastsep == NULL) {
			of->filename = mem_alloc_str(filepath);
		} else {
			of->filename = mem_alloc_str(lastsep+1);
			*lastsep='\0';
			of->filepath = mem_alloc_str(filepath);
		}
		mem_free(filepath);

		return of;
	} else {
		// file not opened
		log_debug("Could not open file: '%s'", filepath);
		mem_free(filepath);
		return NULL;
	}
}

// open a file, checking all the include dirs
static openfile_t *infile_open(const char *filename) {


	// first try with current dir
	char *fullpath = alloc_fullpath(current_file == NULL ? NULL : current_file->filepath, filename);

	// opens file, frees fullpath, returns openfile struct or NULL
	openfile_t *ofile = open_file(fullpath);

	// not found, so try include dirs
	if (ofile == NULL) {

		list_iterator_t *iter = list_iterator(incdirs);
		while (ofile == NULL && list_iterator_has_next(iter)) {
			char *incpath = (char*) list_iterator_next(iter);

			fullpath = alloc_fullpath(incpath, filename);

			ofile = open_file(fullpath);
		}
		list_iterator_free(iter);
	}

	// still not found, so use current dir (if not yet checked if current_file is NULL)
	if (ofile == NULL && current_file != NULL) {

		fullpath = alloc_fullpath(NULL, filename);

		ofile = open_file(fullpath);
	}

	return ofile;
}

void infiles_include(const char *filename) {

	list_add(filestack, current_file);

	current_file = infile_open(filename);
}

// low level close, also releases the allocated attributes (buffer)
// but not the struct itself, as references to it can be held 
static void infiles_close(openfile_t *file) {

	fclose(file->filep);

	mem_free(file->buffer);
	file->buffer = NULL;

	if (current_file == file) {
		current_file = NULL;
	}
}

line_t *infiles_readline() {

	if (infile_iter == NULL) {
		infile_iter = list_iterator(infiles);
	}

	if (current_file != NULL && feof(current_file->filep)) {
		// file completely read, close it
		infiles_close(current_file);
	}

	// first check if we are in an include, and return to parent file
	if (current_file == NULL) {

		current_file = list_pop(filestack);
	}

	// we were not in an include, so try the next top level file	
	if (current_file == NULL) {

		char *filename = list_iterator_next(infile_iter);

		if (filename == NULL) {
			// last input file done
			// end of pass
			return NULL;
		}

		current_file = infile_open(filename);

		if (current_file == NULL) {
			// could not open file
			log_error("Could not open file from all include dirs: %s", filename);
			return NULL;
		}
	}

	char *buffer = current_file->buffer;

	// current_file is not NULL
	char *r = fgets(buffer, current_file->buffer_size, current_file->filep);
	if (r == NULL) {
		// EOF without chars, or error
		if (ferror(current_file->filep)) {
			log_error("Error on reading file %s", current_file->filename);
			infiles_close(current_file);
			return NULL;
		}
		// EOF without chars
		// should not happen

		log_debug("EOF for file %s without chars!", current_file->filename);
		// fake return empty line
		strcpy(current_file->buffer, "\n");
	}

	int len = strlen(buffer);
	if (buffer[len-1] != '\n') {
		if (len >= current_file->buffer_size - 1) {
			// buffer overflow
			log_error("Line buffer overflow on line %d of file %s after %d bytes", 
				current_file->current_line, current_file->filename, current_file->buffer_size);
		}
	} else {
		// overwrite final newline
		buffer[len-1] = 0;
	}


	line.position = mem_alloc(&position_memtype);
	line.position->file = current_file;
	line.position->lineno = current_file->current_line;
	line.line = buffer;

	current_file->current_line++;
	return &line;
}

