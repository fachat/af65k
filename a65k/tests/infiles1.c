
#include <stdbool.h>
#include <string.h>
#include <stdio.h>

#include "log.h"
#include "mem.h"
#include "infiles.h"


void do_test();

int main(int argc, char *argv[]) {


	printf("infiles:\n");

	infiles_module_init();

	infiles_register("testfiles/main.a65");
	infiles_includedir("testfiles/incdir1");

	do_test();

	return 0;
}


void do_test() {

        line_t *line;

	char *tmp = NULL;
	char *ptr = NULL;

        line = infiles_readline();
        while (line != NULL) {




		tmp = mem_alloc_str(line->line);
		if (strlen(tmp) > 110) {
			strcpy(tmp+100, "...");
		}
		printf("got f='%s', l=%03d: %s\n", line->file->filename, line->lineno, tmp);

		char *tok = strtok_r(tmp, " ", &ptr);
		if (tok != NULL && !strcmp("include", tok)) {
			char *incfilename = strtok_r(NULL, " ", &ptr);
			infiles_include(incfilename);
		}
		mem_free(tmp);
		tmp = NULL;

                line = infiles_readline();
        }

}

