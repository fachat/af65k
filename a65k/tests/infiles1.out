infiles:
got f='main.a65', l=001: 
got f='main.a65', l=002: this file is the main test root
got f='main.a65', l=003: 
got f='main.a65', l=004: // include dir test
got f='main.a65', l=005: include inc.i65
DBG:Could not open file: 'testfiles/inc.i65'
got f='inc.i65', l=001: included file
got f='inc.i65', l=002: 
DBG:EOF for file inc.i65 without chars!
got f='inc.i65', l=003: 
got f='main.a65', l=006: 
got f='main.a65', l=007: // file without ln at the end of file
got f='main.a65', l=008: include file_without_ln.a65
got f='file_without_ln.a65', l=001: this file does not have a new line at its end
got f='file_without_ln.a65', l=002: next to last line
got f='file_without_ln.a65', l=003: last line
got f='main.a65', l=009: 
got f='main.a65', l=010: include oversize_line.a65
ERR:Line buffer overflow on line 1 of file oversize_line.a65 after 4096 bytes
got f='oversize_line.a65', l=001: /****************************************************************************    tokenizing a line  ...
ERR:Line buffer overflow on line 2 of file oversize_line.a65 after 4096 bytes
got f='oversize_line.a65', l=002: UAL;				tok->len++;				return true;			default:				// default				return true;			}		} else {			// ">" ...
ERR:Line buffer overflow on line 3 of file oversize_line.a65 after 4096 bytes
got f='oversize_line.a65', l=003: eral Public License    along with this program; if not, write to the Free Software    Foundation, In...
got f='oversize_line.a65', l=004: nfile_iter = list_iterator(infiles);	}	if (current_file != NULL && feof(current_file->filep)) {		// ...
got f='main.a65', l=011: 
got f='main.a65', l=012: last line
DBG:EOF for file main.a65 without chars!
got f='main.a65', l=013: 
